File: Data Structures & Algorithms/house-robber/submission-7.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1: return nums[0]
        if n == 2: return max(nums[0], nums[1]) 

        cache = {0: nums[0], 1: max(nums[0], nums[1])}
        def helper(i):
            if i in cache:
                return cache[i]
            else:
                cache[i] = max(nums[i] + helper(i-2), helper(i-1)) 
                return cache[i]

        return helper(n - 1)

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-7.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        std::unordered_set<int> my_set(nums.begin(), nums.end()); 
        int longest = 0; 

        for (const auto& num : my_set) {
            if (my_set.find(num - 1) == my_set.end()) {
                int length = 1;

                while(my_set.find(num + length) != my_set.end()) {
                    length += 1; 
                }

                longest = std::max(longest, length);
            }
        }
        return longest; 
    }
};